osu.Game/Rulesets/Scoring/HealthProcessor.cs

```diff
@@ -2,45 +2,74 @@
+using System.Linq;
 using osu.Framework.Bindables;
 using osu.Framework.Utils;
 using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Screens.Play;
 
 namespace osu.Game.Rulesets.Scoring
 {
     public abstract partial class HealthProcessor : JudgementProcessor
     {
         /// <summary>
-        /// Invoked when the <see cref="ScoreProcessor"/> is in a failed state.
+        /// Invoked when the <see cref="HealthProcessor"/> is in a failed state.
         /// Return true if the fail was permitted.
         /// </summary>
         public event Func<bool>? Failed;
 
         /// <summary>
-        /// Additional conditions on top of <see cref="CheckDefaultFailCondition"/> that cause a failing state.
+        /// The current selected mods.
         /// </summary>
-        public event Func<HealthProcessor, JudgementResult, bool>? FailConditions;
+        public readonly Bindable<IReadOnlyList<Mod>> Mods = new Bindable<IReadOnlyList<Mod>>(Array.Empty<Mod>());
+
+        /// <summary>
+        /// Mods which can override fail ordered in the correct order for checks.
+        /// </summary>
+        public IApplicableFailOverride[] OrderedFailOverrideMods { get; private set; } = Array.Empty<IApplicableFailOverride>();
 
         /// <summary>
         /// The current health.
         /// </summary>
         public readonly BindableDouble Health = new BindableDouble(1) { MinValue = 0, MaxValue = 1 };
 
         /// <summary>
-        /// Whether this ScoreProcessor has already triggered the failed state.
+        /// Whether this <see cref="HealthProcessor"/> has already triggered the failed state.
         /// </summary>
         public bool HasFailed { get; private set; }
 
+        /// <summary>
+        /// If this <see cref="HealthProcessor"/> is in a failed state due to a mod, this returns the instance of that mod.
+        /// </summary>
+        /// <remarks>
+        /// Used in <see cref="Player"/> to determine whether to perform a restart on failure, if the triggering mod is configured as such.
+        /// </remarks>
+        public Mod? ModTriggeringFailure { get; private set; }
+
         /// <summary>
         /// Immediately triggers a failure for this HealthProcessor.
         /// </summary>
-        public void TriggerFailure()
+        /// <param name="triggeringMod">An optional mod that triggered failure.</param>
+        public void TriggerFailure(Mod? triggeringMod = null)
         {
             if (HasFailed)
                 return;
 
             if (Failed?.Invoke() != false)
                 HasFailed = true;
+
+            if (triggeringMod != null)
+                ModTriggeringFailure = triggeringMod;
+        }
+
+        protected HealthProcessor()
+        {
+            Mods.BindValueChanged(mods =>
+            {
+                OrderedFailOverrideMods = mods.NewValue.OfType<IApplicableFailOverride>().OrderByDescending(m => m.RestartOnFail).ToArray();
+            });
```

Comment:
I'm not sure I understand the implications of this ordering, and I would like to see if it can be removed. It's also not documented anywhere as this seems like very specific logic for `RestartOnFail`.

The way I envision it is as an aggregate condition, in other words (don't use linq if actually implementing this):
```csharp
bool allow = Mods.All(m => m.CheckFail() != Block);
bool force = Mods.Any(m => m.CheckFail() == Force);

if (force || allow)
    // do fail
```

Reply:
The ordering is the crux of this PR, to fix the underlying issue (https://github.com/ppy/osu/issues/28434). It may need better documentation, but reading the issue should explain why it's there, I think.